extensions/eclipse-che-theia-plugin-ext/src/plugin/che-sidecar-file-system.ts

```diff
@@ -0,0 +1,223 @@
+/**********************************************************************
+ * Copyright (c) 2018-2020 Red Hat, Inc.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ***********************************************************************/
+import { CheSideCarFileSystem, FileTypeMain, PLUGIN_RPC_CONTEXT } from '../common/che-protocol';
+import { Stats, lstat, readFile, readdir, stat } from 'fs';
+
+import { RPCProtocol } from '@theia/plugin-ext/lib/common/rpc-protocol';
+import { URI } from 'vscode-uri';
+import { promisify } from 'util';
+
+export namespace SideCarFileSystemProvider {
+  export interface StatAndLink {
+    // The stats of the file. If the file is a symbolic
+    // link, the stats will be of that target file and
+    // not the link itself.
+    // If the file is a symbolic link pointing to a non
+    // existing file, the stat will be of the link and
+    // the `dangling` flag will indicate this.
+    stat: Stats;
+
+    // Will be provided if the resource is a symbolic link
+    // on disk. Use the `dangling` flag to find out if it
+    // points to a resource that does not exist on disk.
+    symbolicLink?: { dangling: boolean };
+  }
+}
+
+export enum FileSystemProviderErrorCode {
+  FileExists = 'EntryExists',
+  FileNotFound = 'EntryNotFound',
+  FileNotADirectory = 'EntryNotADirectory',
+  FileIsADirectory = 'EntryIsADirectory',
+  FileExceedsMemoryLimit = 'EntryExceedsMemoryLimit',
+  FileTooLarge = 'EntryTooLarge',
+  NoPermissions = 'NoPermissions',
+  Unavailable = 'Unavailable',
+  Unknown = 'Unknown',
+}
+
+export class FileSystemProviderError extends Error {
+  constructor(message: string, public readonly code: FileSystemProviderErrorCode) {
+    super(message);
+    Object.setPrototypeOf(this, FileSystemProviderError.prototype);
+  }
+}
+
+export function markAsFileSystemProviderError(error: Error, code: FileSystemProviderErrorCode): Error {
+  error.name = code ? `${code} (FileSystemError)` : 'FileSystemError';
+
+  return error;
+}
+
+export function createFileSystemProviderError(
+  error: Error | string,
+  code: FileSystemProviderErrorCode
+): FileSystemProviderError {
+  const providerError = new FileSystemProviderError(error.toString(), code);
+  markAsFileSystemProviderError(providerError, code);
+
+  return providerError;
+}
+
+export class CheSideCarFileSystemImpl implements CheSideCarFileSystem {
```

Comment:
Have you considered using the stuff from `remote-files-system-provider.ts`? I know it's based on injection, but that could be worked around.

Reply:
In particular, we could use a DiskFileSystemProvider for the implementation of the file system and just mangle the URIs.

Reply:
Not sure that it will get a proper benefit. At this time we need at most two operations `fs.stat` and `fs.readFile`. Because DiskFileSystemProvider has an implementation for the streamed file reading and writing, not sure if it is possible to implement via RPC calls. But when we'll need more file system operations, we can partially reuse DiskFileSystemProvider.